processor/transformprocessor/internal/metrics/func_extract_sum_metric.go

```diff
@@ -0,0 +1,108 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+package metrics // import "github.com/open-telemetry/opentelemetry-collector-contrib/processor/transformprocessor/internal/metrics"
+
+import (
+	"context"
+	"fmt"
+
+	"go.opentelemetry.io/collector/pdata/pcommon"
+	"go.opentelemetry.io/collector/pdata/pmetric"
+
+	"github.com/open-telemetry/opentelemetry-collector-contrib/pkg/ottl"
+	"github.com/open-telemetry/opentelemetry-collector-contrib/pkg/ottl/contexts/ottlmetric"
+)
+
+type extractSumMetricArguments struct {
+	Monotonic bool      `ottlarg:"0"`
+	AggTemp   ottl.Enum `ottlarg:"1"`
+}
+
+func newExtractSumMetricFactory() ottl.Factory[ottlmetric.TransformContext] {
+	return ottl.NewFactory("extract_sum_metric", &extractSumMetricArguments{}, createExtractSumMetricFunction)
+}
+
+func createExtractSumMetricFunction(_ ottl.FunctionContext, oArgs ottl.Arguments) (ottl.ExprFunc[ottlmetric.TransformContext], error) {
+	args, ok := oArgs.(*extractSumMetricArguments)
+
+	if !ok {
+		return nil, fmt.Errorf("extractSumMetricFactory args must be of type *extractSumMetricArguments")
+	}
+
+	return extractSumMetric(args.Monotonic, args.AggTemp)
+}
+
+// this interface helps unify the logic for extracting data from different histogram types
+// all supported metric types' datapoints implement it
+type SumCountDataPoint interface {
+	Attributes() pcommon.Map
+	Sum() float64
+	Count() uint64
+	StartTimestamp() pcommon.Timestamp
+	Timestamp() pcommon.Timestamp
+}
+
+func extractSumMetric(monotonic bool, aggTempEnum ottl.Enum) (ottl.ExprFunc[ottlmetric.TransformContext], error) {
+	aggTemp := pmetric.AggregationTemporality(aggTempEnum)
+	switch aggTemp {
+	case pmetric.AggregationTemporalityDelta, pmetric.AggregationTemporalityCumulative:
+	default:
+		return nil, fmt.Errorf("unknown aggregation temporality: %s", aggTemp.String())
+	}
+
+	return func(_ context.Context, tCtx ottlmetric.TransformContext) (interface{}, error) {
+		metric := tCtx.GetMetric()
+		invalidMetricTypeError := fmt.Errorf("extract_sum_metric requires an input metric of type Histogram, ExponentialHistogram or Summary, got %s", metric.Type())
+
+		switch metric.Type() {
+		case pmetric.MetricTypeHistogram, pmetric.MetricTypeExponentialHistogram, pmetric.MetricTypeSummary:
+		default:
+			return nil, invalidMetricTypeError
+		}
+
+		sumMetric := tCtx.GetMetrics().AppendEmpty()
+		sumMetric.SetDescription(metric.Description())
+		sumMetric.SetName(metric.Name() + "_sum")
+		sumMetric.SetUnit(metric.Unit())
+		sumMetric.SetEmptySum().SetAggregationTemporality(aggTemp)
```

Comment:
Sorry to be so persistent, but I really think we should only allow setting the aggregation temporality for summaries and make setting it for histograms a no-op (and then an error at some point once it is an optional argument). Unless the temporality of the incoming data has been incorrectly set, switching the temporality of Histograms without performing the conversion will produce bad data. I expect this won't be the case for the vast majority of users, so passing this argument will only go to create more issues than it solves. Users who truly want to change the declared aggregation temporality without performing a conversion can still use the `set` function.

Once `aggTempEnum` is marked as optional, we should default it to `cumulative` for Summaries. Summaries are intended for [Prometheus and OpenMetrics compatability](https://github.com/open-telemetry/opentelemetry-proto/blob/main/opentelemetry/proto/metrics/v1/metrics.proto#L233) per the proto. Prometheus uses cumulative aggregation temporality to my knowledge, and the `sum` value in Summaries is declared to be a [cumulative value](https://github.com/OpenObservability/OpenMetrics/blob/main/specification/OpenMetrics.md#summary) per the OpenMetrics specification ("Count and Sum values are counters so MUST NOT be NaN or negative"). While Prometheus and OpenMetrics will produce the majority of Summary metrics, we don't know where the input metrics came from, so I think it still makes sense to allow the user to configure this even if it's an uncommon case.

Reply:
I agree that quietly ignoring it isn't ideal, and that we should trust users to make the right decisions about their data. My specific concern here is that allowing users to set the aggregation temporality seems like a needless sharp edge when there's only one right option, and we already know what it is. Put another way, I don't feel like the user has any meaningful autonomy here.

In place of a temporary no-op for setting the aggregation temporality for histograms, I'd propose two alternatives:
1. Have `extract_sum_from_histogram(monotonic)` and `extract_sum_from_summary(monotonic, agg_temp)`. This way it's clear up-front what the options are from the signatures.
2. Exclude aggregation temporality as an argument for now, and copy it from the input metric for histograms, or set it to cumulative for summaries. Keep the behavior we've discussed once optional parameters land. Since the parameter is optional, this would be a non-breaking change.

Reply:
I've changed this here: https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/24368/commits/923c2e09dcde239ef3cd71661b3b69f46cca0f1b. If we decide to go a different way, it's easy to just revert the commit.